### PR TITLE
integrate get_history_balance_er20

### PR DESCRIPTION
**Pull request by malikfassi**

### CODE REVIEW

File: src/main/scala/co/ledger/wallet/daemon/services/AccountsService.scala [hunk 1]
```diff
@@ -72,6 +73,21 @@ class AccountsService @Inject()(daemonCache: DaemonCache) extends DaemonService
...
 
+  def getTokenCoreAccount(tokenAccountInfo: TokenAccountInfo): Future[core.ERC20LikeAccount] =
+    daemonCache.withAccount(tokenAccountInfo.accountInfo)(_.erc20Account(tokenAccountInfo.tokenAddress).liftTo[Future])
+
+  def getTokenCoreAccountBalanceHistory(tokenAccountInfo: TokenAccountInfo, start: String, end: String, period: core.TimePeriod): Future[ArrayList[core.BigInt]] = {
+    val format = new java.text.SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ")
+    val startDate = format.parse(start)
+    val endDate = format.parse(end)
+
+     for {
```
Comment (luchenyuxx) [inline src/main/scala/co/ledger/wallet/daemon/services/AccountsService.scala]:
```suggestion
     getTokenCoreAccount(tokenAccountInfo).map(_.getBalanceHistoryFor(startDate, endDate, period))
```

File: src/main/scala/co/ledger/wallet/daemon/controllers/AccountsController.scala [hunk 2]
```diff
@@ -162,6 +163,15 @@ class AccountsController @Inject()(accountsService: AccountsService) extends Con
         accountsService.getTokenAccount(request.tokenAccountInfo)
       }
 
+      // Return the balances and operation counts history of token accounts in the order of the starting time to the end time.
+      get("/tokens/:token_address/history") { request: TokenHistoryRequest =>
+        info(s"Get history $request")
+        for {
```
Comment (luchenyuxx) [inline src/main/scala/co/ledger/wallet/daemon/controllers/AccountsController.scala]:
```suggestion
        accountsService.getTokenCoreAccountBalanceHistory(request.tokenAccountInfo, request.startDate, request.endDate, request.timePeriod).map(TokenHistoryResponse(_))
```

File: src/main/scala/co/ledger/wallet/daemon/services/AccountsService.scala [hunk 3]
```diff
@@ -72,6 +73,21 @@ class AccountsService @Inject()(daemonCache: DaemonCache) extends DaemonService
   def getTokenAccount(tokenAccountInfo: TokenAccountInfo): Future[ERC20AccountView] =
     daemonCache.withAccount(tokenAccountInfo.accountInfo)(_.erc20Account(tokenAccountInfo.tokenAddress).map(ERC20AccountView(_)).liftTo[Future])
 
+  def getTokenCoreAccount(tokenAccountInfo: TokenAccountInfo): Future[core.ERC20LikeAccount] =
+    daemonCache.withAccount(tokenAccountInfo.accountInfo)(_.erc20Account(tokenAccountInfo.tokenAddress).liftTo[Future])
+
+  def getTokenCoreAccountBalanceHistory(tokenAccountInfo: TokenAccountInfo, start: String, end: String, period: core.TimePeriod): Future[ArrayList[core.BigInt]] = {
```
Comment (luchenyuxx) [inline src/main/scala/co/ledger/wallet/daemon/services/AccountsService.scala]:
```suggestion
  def getTokenCoreAccountBalanceHistory(tokenAccountInfo: TokenAccountInfo, start: Date, end: Date, period: core.TimePeriod): Future[ArrayList[core.BigInt]] = {
```

File: src/main/scala/co/ledger/wallet/daemon/services/AccountsService.scala [hunk 4]
```diff
@@ -72,6 +73,21 @@ class AccountsService @Inject()(daemonCache: DaemonCache) extends DaemonService
...
+    val format = new java.text.SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ")
```
Comment (luchenyuxx) [inline src/main/scala/co/ledger/wallet/daemon/services/AccountsService.scala]:
```suggestion
```

File: src/main/scala/co/ledger/wallet/daemon/services/AccountsService.scala [hunk 5]
```diff
@@ -72,6 +73,21 @@ class AccountsService @Inject()(daemonCache: DaemonCache) extends DaemonService
...
+    val startDate = format.parse(start)
```
Comment (luchenyuxx) [inline src/main/scala/co/ledger/wallet/daemon/services/AccountsService.scala]:
```suggestion
```

File: src/main/scala/co/ledger/wallet/daemon/services/AccountsService.scala [hunk 6]
```diff
@@ -72,6 +73,21 @@ class AccountsService @Inject()(daemonCache: DaemonCache) extends DaemonService
...
+    val endDate = format.parse(end)
```
Comment (luchenyuxx) [inline src/main/scala/co/ledger/wallet/daemon/services/AccountsService.scala]:
```suggestion
```

File: src/main/scala/co/ledger/wallet/daemon/controllers/AccountsController.scala [hunk 7]
```diff
@@ -168,6 +169,16 @@ class AccountsController @Inject()(accountsService: AccountsService) extends Con
         accountsService.getTokenAccount(request.tokenAccountInfo)
       }
 
+      // Return the balances and operation counts history of token accounts in the order of the starting time to the end time.
+      get("/tokens/:token_address/history") { request: TokenHistoryRequest =>
+        for {
+          balances <- accountsService.getTokenCoreAccountBalanceHistory(request.tokenAccountInfo, request.startDate, request.endDate, request.timePeriod)
+          accountOpt <- accountsService.getAccount(request.tokenAccountInfo.accountInfo)
+          account = accountOpt.getOrElse(throw AccountNotFoundException(request.account_index))
```
Comment (aymeric-ledger) [inline src/main/scala/co/ledger/wallet/daemon/controllers/AccountsController.scala]: compose using `Future` instead of throwing like this.

**Comment by phaazon** [comment]
@malikfassi do you still have weird issues with the `BigInt`? Also, I’m pinging @KhalilBellakrid as he wants to take ownership on the Ledger Core side about this.

**Comment by malikfassi** [comment]
@phaazon This has been fixed. It was indeed an issue with types. I wanted everything to work before announcing it.

**Comment by phaazon** [comment]
@malikfassi amazing, thanks! Do you have hindsight about what was causing the problem with those specific types? :)

File: src/main/scala/co/ledger/wallet/daemon/models/Account.scala [hunk 8]
```diff
@@ -263,6 +266,18 @@ object Account extends Logging {
     }
   }
 
+  // TODO: refactor this part once lib-core provides the feature
```
Comment (malikfassi) [inline src/main/scala/co/ledger/wallet/daemon/models/Account.scala]: @KhalilBellakrid Is it something feasible in lib-core ?

**Pull request merged**